app/src/main/java/io/github/trojan_gfw/igniter/servers/presenter/ServerListPresenter.java

```diff
@@ -43,6 +46,51 @@ public void displayImportFileDescription() {
         mView.showImportFileDescription();
     }
 
+    @Override
+    public void exportServerListToFile() {
+        Threads.instance().runOnWorkThread(new Task() {
+            @Override
+            public void onRun() {
+                exportConfigFile();
+            }
+        });
+    }
+
+    private void exportConfigFile() {
+        boolean success = false;
+        try {
+            String exportContent = getExportContent();
+            success = TrojanHelper.writeStringToFile(exportContent, Globals.getIgniterExportPath());
```

Comment:
Noticed that you attempt to export file on external storage, which requires permission `WRITE_EXTERNAL_STORAGE`. You have to check whether the permission is granted using `ContextCompat.checkSelfPermission(Context,String)` and make permission request if necessary.
BTW, do not forget to declare `WRITE_EXTERNAL_STORAGE` permission in `AndroidManifest.xml`.

Reply:
On Android 4.4 (API level 19) or higher, your app doesn't need to request any storage-related permissions to access app-specific directories within external storage. The `minimumSdkVersion` in build.gradle is 21, it'll work ok.

Reply:
> On Android 4.4 (API level 19) or higher, your app doesn't need to request any storage-related permissions to access app-specific directories within external storage. The `minimumSdkVersion` in build.gradle is 21, it'll work ok.

My bad, I forgot the api restriction, you are right.